org.eclipse.xtext.util/src/org/eclipse/xtext/util/concurrent/IReadAccess.java

```diff
@@ -126,6 +144,25 @@
 			IUnitOfWork<Result, State> work,
 			Supplier<? extends Result> defaultResult
 		) {
+			// Some implementations rely on the type of {@code work}
+			if (work instanceof CancelableUnitOfWork<?, ?>) {
+				return priorityReadOnly(new CancelableUnitOfWork<Result, State>() {
```

Comment:
Can we extract the anonymous class into a named one since the code is the same for cases?